crates/worker/src/partition_processor_manager/spawn_processor_task.rs

```diff
@@ -233,3 +174,158 @@ impl SpawnPartitionProcessorTask {
+    let snapshot = match snapshot_repository {
+        Some(repository) => {
+            debug!("Looking for partition snapshot from which to bootstrap partition store");
+            // todo(pavel): pass target LSN to repository
+            repository.get_latest(partition_id).await?
+        }
+        None => {
+            debug!("No snapshot repository configured");
+            None
+        }
+    };
+
+    match (snapshot, fast_forward_lsn) {
+        (None, None) => {
+            info!("No snapshot found to bootstrap partition, creating new store");
+            Ok(partition_store_manager
+                .open_partition_store(
+                    partition_id,
+                    key_range,
+                    OpenMode::CreateIfMissing,
+                    &options.storage.rocksdb,
+                )
+                .await?)
+        }
+        (Some(snapshot), None) => {
+            info!(partition_id = %partition_id, "Found partition snapshot, restoring it");
+            Ok(import_snapshot(
+                partition_id,
+                key_range,
+                snapshot,
+                partition_store_manager,
+                options,
+            )
+            .await?)
+        }
+        (Some(snapshot), Some(fast_forward_lsn))
+            if snapshot.min_applied_lsn >= fast_forward_lsn =>
+        {
+            // We trust that the fast_forward_lsn is greater than the locally applied LSN.
+            info!(
+                latest_snapshot_lsn = ?snapshot.min_applied_lsn,
+                ?fast_forward_lsn,
+                "Found snapshot with LSN >= target LSN, dropping local partition store state",
+            );
+            partition_store_manager.drop_partition(partition_id).await;
+            Ok(import_snapshot(
+                partition_id,
+                key_range,
+                snapshot,
+                partition_store_manager,
+                options,
+            )
+            .await?)
+        }
+        (maybe_snapshot, Some(fast_forward_lsn)) => {
+            // Play it safe and keep the partition store intact; we can't do much else at this
+            // point. We'll likely halt again as soon as the processor starts up.
+            if let Some(snapshot) = maybe_snapshot {
+                warn!(
+                    ?snapshot.min_applied_lsn,
+                    ?fast_forward_lsn,
+                    "Latest available snapshot is older than the the fast-forward target LSN!",
+                );
+            } else {
+                warn!(
+                    ?fast_forward_lsn,
+                    "A fast-forward target LSN is set, but no snapshot available for partition!",
+                );
+            }
+
+            // We expect the processor startup attempt will fail, avoid spinning too fast.
+            tokio::time::sleep(Duration::from_millis(
+                10_000 + rand::random::<u64>() % 10_000,
+            ))
```

Comment:
would RetryPolicy and its internal jitter logic work for you here?

Reply:
I remembered why I didn't want to tackle this just yet - right now the way to get consecutive retry decisions is to get an iterator. I want to introduce an alternative API to `RetryPolicy` which will make this more suitable for use cases like this one, but let me rather do that as a follow up PR!